model-datastructure/src/jvmMain/kotlin/org/modelix/model/InMemoryModel.kt

```diff
@@ -0,0 +1,293 @@
+/*
+ * Copyright (c) 2024.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.modelix.model
+
+import gnu.trove.map.TLongObjectMap
+import gnu.trove.map.hash.TLongObjectHashMap
+import org.modelix.model.api.ConceptReference
+import org.modelix.model.api.IBranch
+import org.modelix.model.api.IConcept
+import org.modelix.model.api.IConceptReference
+import org.modelix.model.api.ILanguageRepository
+import org.modelix.model.api.INode
+import org.modelix.model.api.INodeReference
+import org.modelix.model.api.ITree
+import org.modelix.model.api.NodeReference
+import org.modelix.model.api.PNodeReference
+import org.modelix.model.api.resolveInCurrentContext
+import org.modelix.model.area.IArea
+import org.modelix.model.area.IAreaListener
+import org.modelix.model.area.IAreaReference
+import org.modelix.model.lazy.CLTree
+import org.modelix.model.lazy.KVEntryReference
+import org.modelix.model.lazy.NonCachingObjectStore
+import org.modelix.model.persistent.CPHamtNode
+import org.modelix.model.persistent.CPNode
+import org.modelix.model.persistent.CPNodeRef
+import kotlin.system.measureTimeMillis
+import kotlin.time.Duration.Companion.milliseconds
+import kotlin.time.DurationUnit
+
+private val LOG = mu.KotlinLogging.logger { }
+
+class IncrementalInMemoryModel {
+    private var lastModel: InMemoryModel? = null
+
+    @Synchronized
+    fun getModel(tree: CLTree): InMemoryModel {
+        val reusable = lastModel?.takeIf { it.branchId == tree.getId() }
+        val newModel = if (reusable == null) {
+            InMemoryModel.load(tree)
```

Comment:
Isn't the `tree.getId()` actually the (internal) repository ID?

Does this mean, that we evict and reload the `lastModel` continuously, when the server servers two repositories by ModelQL?

(Still better than before, but I want to understand the limitations.)

Reply:
I didn't want to have an unlimited amount of models in memory, but I changed it.